js/autotheme.js

```diff
@@ -0,0 +1,174 @@
+/*global Color, ColorThief, Defer, EventDispatcher, MozActivity */
+
+'use strict';
+
+(function(exports) {
+  // is autotheming active
+  var active;
+
+  function loadImage(src) {
+    var defer = new Defer();
+
+    var img = new Image();
+    img.src = src;
+
+    if (img.complete) {
+      defer.resolve(img);
+    } else {
+      img.onload = () => defer.resolve(img);
+      img.onerror = defer.reject;
+    }
+
+    return defer.promise;
+  }
+
+  function blobFromURL(src) {
+    var defer = new Defer();
+
+    var xhr = new XMLHttpRequest();
+    xhr.addEventListener('load', () => {
+      defer.resolve(xhr.response);
+    });
+    xhr.addEventListener('error', defer.reject);
+
+    xhr.open('GET', src);
+    xhr.responseType = 'blob';
+    xhr.send();
+
+    return defer.promise;
+  }
+
+  var AutoTheme = {
+    get active() {
+      return active;
+    },
+    set active(bool) {
+      active = bool;
+      document.body.classList.toggle('has-autotheme', bool);
+    },
+    elts: {
+      commandCreate: document.querySelectorAll('.autotheme-command-create'),
+      commandCancel: document.querySelectorAll('.autotheme-command-cancel')
+    },
+
+    handleEvent(e) {
+      switch(e.type) {
+        case 'click':
+          this.pickImage()
+            .then(this.loadBlob.bind(this))
+            .then(this.storePalette.bind(this));
+          break;
+      }
+    },
+
+    /* called from activity */
+    load(blob) {
+      return this.loadBlob(blob)
+        .then(this.storePalette.bind(this));
+    },
+
+    pickImage() {
+      var defer = new Defer();
+
+      if (window.MozActivity) {
+        var activity = new MozActivity({
+          name: 'pick',
+          data: { type: 'image/*' }
+        });
+
+        activity.onsuccess = () => defer.resolve(activity.result.blob);
+        activity.onerror = defer.reject;
+      } else {
+        defer.resolve(blobFromURL('img/default_image.jpg'));
+      }
+
+      return defer.promise;
+    },
+
+    loadBlob(blob) {
```

Comment:
this one really loads a blob and returns a palette. Maybe I should call this blobToPalette?